internal/scheduler/reports.go

```diff
@@ -0,0 +1,277 @@
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/google/uuid"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/openconfig/goyang/pkg/indent"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/G-Research/armada/internal/common/armadaerrors"
+	"github.com/G-Research/armada/internal/scheduler/schedulerobjects"
+	"github.com/G-Research/armada/pkg/api"
+	"github.com/G-Research/armada/pkg/armadaevents"
+)
+
+// SchedulingReportsRepository stores reports on the most recent scheduling attempts.
+type SchedulingReportsRepository struct {
+	// Scheduling reports for the jobs that were most recently attempted to be scheduled.
+	MostRecentJobSchedulingReports *lru.Cache
+	// Scheduling reports for the most recently seen queues.
+	MostRecentQueueSchedulingReports *lru.Cache
+}
+
+func NewSchedulingReportsRepository(maxQueueSchedulingReports, maxJobSchedulingReports int) *SchedulingReportsRepository {
+	mostRecentJobSchedulingReports, err := lru.New(maxJobSchedulingReports)
+	if err != nil {
+		panic(errors.WithStack(err))
+	}
+	mostRecentQueueSchedulingReports, err := lru.New(maxQueueSchedulingReports)
+	if err != nil {
+		panic(errors.WithStack(err))
+	}
+	return &SchedulingReportsRepository{
+		MostRecentJobSchedulingReports:   mostRecentJobSchedulingReports,
+		MostRecentQueueSchedulingReports: mostRecentQueueSchedulingReports,
+	}
+}
+
+func (repo *SchedulingReportsRepository) GetQueueReport(ctx context.Context, queue *schedulerobjects.Queue) (*schedulerobjects.QueueReport, error) {
+	report, ok := repo.GetQueueSchedulingReport(queue.Name)
+	if !ok {
+		return nil, &armadaerrors.ErrNotFound{
+			Type:    "QueueSchedulingReport",
+			Value:   queue.Name,
+			Message: "this queue has not been considered for scheduling recently",
+		}
+	}
+	return &schedulerobjects.QueueReport{
+		Report: report.String(),
+	}, nil
+}
+
+func (repo *SchedulingReportsRepository) GetJobReport(ctx context.Context, jobId *schedulerobjects.JobId) (*schedulerobjects.JobReport, error) {
+	jobUuid, err := uuidFromUlidString(jobId.Id)
+	if err != nil {
+		return nil, err
+	}
+	report, ok := repo.GetJobSchedulingReport(jobUuid)
+	if !ok {
+		return nil, &armadaerrors.ErrNotFound{
```

Comment:
should this (and the one above) be errors?

Reply:
Yes. So that gRPC error codes are set correctly.